### PR TITLE
Add runtime function to query Kokkos initialization settings

### PR DESCRIPTION
**Pull request by dalg24**
Tentative resolution for #5843 

Adding `Kokkos::initialization_settings()` runtime function that returns the settings including the environment variable parsing.

### CODE REVIEW

**Comment by crtrott** [comment]
Yeah looks fine.

File: core/src/impl/Kokkos_Core.cpp [hunk 1]
```diff
@@ -767,7 +770,8 @@ void post_initialize_internal(const Kokkos::InitializationSettings& settings) {
   Kokkos::Tools::InitArguments tools_init_arguments;
   combine(tools_init_arguments, settings);
   initialize_profiling(tools_init_arguments);
-  g_is_initialized = true;
+  g_is_initialized          = true;
+  g_initialization_settings = settings;
```
Comment (kmorel) [inline core/src/impl/Kokkos_Core.cpp:774]: At this point, does `settings` contain information about settings that are implicitly set? For example, if the user doesn't set num threads (for a CPU device), does _settings_ have the number of threads that will actually be used?

Comment (dalg24) [reply core/src/impl/Kokkos_Core.cpp:774]: No.  It contains the settings resolved from the argument passed to `Kokkos::initialize` and the environment variables.

Comment (kmorel) [reply core/src/impl/Kokkos_Core.cpp:774]: Is there a way to fill it with that information? That is something we would like to capture if possible.

Comment (dalg24) [reply core/src/impl/Kokkos_Core.cpp:774]: Can you elaborate on why the version proposed is not sufficient?
If `device_id` is set, you will get that device or Kokkos will call abort if it is not able to.  We will not just decide to pick another one.  Similarly, if `num_threads` is specified, then Kokkos will honor that choice.

Comment (kmorel) [reply core/src/impl/Kokkos_Core.cpp:774]: The point is to get a hint about what kokkos is doing when something is _not_ set. If I don't set `num_threads`, how many threads will Kokkos use? The answer is not obvious. Will it use the number of physical threads, or will it include hyperthreads? Did a job scheduler partition threads in a way that reduces them per process?

Likewise, if I _don't_ set `device_id`, I might want to get an indication of which id Kokkos picked. Did the job scheduler properly allocate different devices for each process on the same node, or did I dork something up with the scheduling flags?

Comment (dalg24) [reply core/src/impl/Kokkos_Core.cpp:774]: OK.  I would explore another direction for this.  I don't like the idea of our returning `InitializationSettings` for that purpose.  Let's take back the discussion to #5843

I will keep that draft open a bit longer to see if there is interest in that facility.

**Comment by dalg24** [comment]
Closing as there does not seem to be much appetite for this facility.  Feel free to comment here if that changes and we can revisit as needed.

**Pull request closed without merging**